Introduce rebar_uri to avoid deprecation warnings on OTP 23 [master]

michaelklishin opened the pull request:
See #2191 for the background.

The new module attempts to act as a shim between `http_uri` (deprecated in OTP 23) and `uri_string` (available as of OTP 21+) with as little wheel reinvention as possible.

michaelklishin commented on test/rebar_utils_SUITE.erl:
```diff
@@ -322,6 +322,7 @@ is_list_of_strings(_Config) ->
 
 url_append_path(_Config) ->
-    ?assertEqual({ok, "https://repo.hex.pm:443/repos/org"}, rebar_utils:url_append_path("https://repo.hex.pm", "/repos/org")),
-    ?assertEqual({ok, "https://repo.hex.pm:443/repos/org?foo=bar"}, rebar_utils:url_append_path("https://repo.hex.pm",
```
I think this test case is wrong. `/repos/org?key=value` is not a path segment. The new module will correctly encode the question mark but previous string concatenation one would leave it unencoded. So the test was updated to use a more sensible path "extension".

ferd replied:
Agreed. Hopefully we just didn't have a thing that relied on that behaviour in a critical way.

ferd commented on test/rebar_uri_SUITE.erl:
```diff
@@ -0,0 +1,32 @@
+    #{scheme := Scheme, host := Host, path := Path} = rebar_uri:parse("https://repo.hex.pm"),
+    ?assertEqual("https", Scheme),
+    ?assertEqual("repo.hex.pm", Host),
+    ?assert(lists:member(Path, ["", "/"])),
```
is this because the behaviour changes based on the version used?

michaelklishin replied:
Correct. `uri_string:parse/1` is not a 100% equivalent to `http_uri:parse/1`.

michaelklishin commented on src/r3_hex_api.erl:
```diff
@@ -28,12 +28,17 @@ put(Config, Path, Body) ->
 delete(Config, Path) ->
     request(Config, delete, Path, undefined).
 
+%% OTP 21+
```
Noted.

michaelklishin commented on src/rebar_utils.erl:
```diff
@@ -935,14 +939,30 @@ normalise_proxy(Scheme, URI) ->
+         _ ->
+             error
+     end.
+-endif.
```
We can. I tried to introduce as few changes as possible.

michaelklishin commented:
I have submitted https://github.com/hexpm/hex_core/pull/82.

michaelklishin commented on src/rebar_uri.erl:
```diff
@@ -0,0 +1,44 @@
+                %% string:slice/2 isn't available in OTP <= 19.
+                query => case Query of
+                           [] -> "";
+                           _  -> string:substr(Query, 2)
```
This function is marked as obsolete in the docs but `string:slice/{2,3}` is not available in OTP 18 and 19. `string:slice/{2,3}` is Unicode-aware as the doc examples demonstrate. I think for this specific case it's fine (we are chopping the leading `?` if any) and the best we can do for old OTP versions.

michaelklishin commented:
I moved `append_path/2` to `rebar_uri` and addressed a few bits of feedback. Do we depend on https://github.com/hexpm/hex_core/pull/82 being merged first? I received no response there so far, likely due to the holidays.

ferd commented:
I think it would make sense to have the hex_core PR come in first. However if we want to move this one forwards, I'd say that we would have to just exclude the change to vendored files; this does not get rid of all warnings, but moves the overall state forwards by removing most warnings.

michaelklishin commented:
Removing only some warnings won't make Rebar build on OTP 23 with "warnings as errors" on.

ferd commented:
It won't, but it's not worse than the current state of this PR while we wait for the other one to move forwards. It just reduces the risk of clashing changes being done in the meanwhile.

michaelklishin commented:
@ferd https://github.com/hexpm/hex_core/pull/82 is in, besides looking for any potential differences in the vendored file, is there anything I should do to update this PR?

ferd commented:
I don't think so. I believe @starbelly was working on integrating the hex-core stuff. They changed how they validate packages, which turns to have an impact on the lockfiles, which requires more complex changes on that end as well.

michaelklishin commented:
So this can go in now then? Or should I just wait? :)

starbelly commented:
This should be closed in favor of #2213

michaelklishin commented:
This involves changes beyond #2213.

michaelklishin commented:
I will revert the `src/r3_hex_api.erl` changes.

ferd commented:
yeah only need to drop the vendored file changes, then I'll merge this, and #2213 will complete the objective for this PR. Next release will build warning-free regardless.

starbelly commented:
@michaelklishin  

> This involves changes beyond #2213.

Yup, my bad.

michaelklishin commented:
@ferd should be ready to go. I ended up undoing `r3_hex_api` changes without `rebase -i`, hopefully this would be acceptable as is.
